exercises/chapter7/src/Data/AddressBook/Validation.purs

```diff
@@ -10,74 +10,129 @@ import Data.Traversable (traverse)
+lengthIs _     _   _     = pure unit
+-- ANCHOR_END: lengthIs
 
+-- ANCHOR: phoneNumberRegex
 phoneNumberRegex :: Regex
 phoneNumberRegex =
   unsafePartial case regex "^\\d{3}-\\d{3}-\\d{4}$" noFlags of
     Right r -> r
+-- ANCHOR_END: phoneNumberRegex
 
+-- ANCHOR: matches
 matches :: String -> Regex -> String -> V Errors Unit
-matches _ regex value
-  | test regex value = pure unit
-
-matches field _ _ = invalid [ "Field '" <> field <> "' did not match the required format" ]
+matches _     regex value | test regex value =
+  pure unit
+matches field _     _     =
+  invalid [ "Field '" <> field <> "' did not match the required format" ]
+-- ANCHOR_END: matches
 
+-- ANCHOR: validateAddress
 validateAddress :: Address -> V Errors Address
 validateAddress a =
-  address <$> (nonEmpty "Street" a.street *> pure a.street)
-    <*> (nonEmpty "City" a.city *> pure a.city)
-    <*> (lengthIs "State" 2 a.state *> pure a.state)
+  address <$> (nonEmpty "Street"  a.street *> pure a.street)
+          <*> (nonEmpty "City"    a.city   *> pure a.city)
+          <*> (lengthIs "State" 2 a.state  *> pure a.state)
+-- ANCHOR_END: validateAddress
 
+-- ANCHOR: validateAddressAdo
 validateAddressAdo :: Address -> V Errors Address
 validateAddressAdo a = ado
-  street <- (nonEmpty "Street" a.street *> pure a.street)
-  city <- (nonEmpty "City" a.city *> pure a.city)
-  state <- (lengthIs "State" 2 a.state *> pure a.state)
+  street <- (nonEmpty "Street"  a.street *> pure a.street)
+  city   <- (nonEmpty "City"    a.city   *> pure a.city)
+  state  <- (lengthIs "State" 2 a.state  *> pure a.state)
   in address street city state
+-- ANCHOR_END: validateAddressAdo
 
+-- ANCHOR: validatePhoneNumber
 validatePhoneNumber :: PhoneNumber -> V Errors PhoneNumber
 validatePhoneNumber pn =
   phoneNumber <$> pure pn."type"
-    <*> (matches "Number" phoneNumberRegex pn.number *> pure pn.number)
+              <*> (matches "Number" phoneNumberRegex pn.number *> pure pn.number)
+-- ANCHOR_END: validatePhoneNumber
 
+-- ANCHOR: validatePhoneNumberAdo
 validatePhoneNumberAdo :: PhoneNumber -> V Errors PhoneNumber
 validatePhoneNumberAdo pn = ado
-  tpe <- pure pn."type"
+  tpe    <- pure pn."type"
   number <- (matches "Number" phoneNumberRegex pn.number *> pure pn.number)
   in phoneNumber tpe number
+-- ANCHOR_END: validatePhoneNumberAdo
 
+-- ANCHOR: validatePerson
 validatePerson :: Person -> V Errors Person
 validatePerson p =
   person <$> (nonEmpty "First Name" p.firstName *> pure p.firstName)
-    <*> (nonEmpty "Last Name" p.lastName *> pure p.lastName)
-    <*> validateAddress p.homeAddress
-    <*> (arrayNonEmpty "Phone Numbers" p.phones *> traverse validatePhoneNumber p.phones)
+         <*> (nonEmpty "Last Name" p.lastName   *> pure p.lastName)
+         <*> validateAddress p.homeAddress
+         <*> (arrayNonEmpty "Phone Numbers" p.phones *>
+              traverse validatePhoneNumber p.phones)
+-- ANCHOR_END: validatePerson
```

Comment:
Follow-up to previous collision comment. This version of `validatePerson` that's actually used.